pkg/cli/cliexecuter.go

```diff
@@ -1,97 +1,268 @@
 package cli
 
 import (
-	"bytes"
+	"context"
 	"errors"
+	"fmt"
+	"os/exec"
+	"strings"
+	"sync"
+	"time"
 
 	"github.com/go-logr/logr"
 
-	corev1 "k8s.io/api/core/v1"
-	"k8s.io/client-go/kubernetes"
-	"k8s.io/client-go/kubernetes/scheme"
-	restclient "k8s.io/client-go/rest"
-	"k8s.io/client-go/tools/remotecommand"
+	apiErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/util/retry"
 	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/medik8s/fence-agents-remediation/api/v1alpha1"
+	"github.com/medik8s/fence-agents-remediation/pkg/utils"
+)
+
+const (
+	FenceAgentContextCanceledMessage = "fence agent context canceled. Nothing to do"
+	FenceAgentContextTimedOutMessage = "fence agent context timed out"
+	FenceAgentRetryErrorMessage      = "fence agent retry error"
 )
 
-type Executer interface {
-	Execute(pod *corev1.Pod, command []string) (stdout string, stderr string, err error)
+type routine struct {
+	cancel context.CancelFunc
 }
 
-type executer struct {
-	log       logr.Logger
-	config    *restclient.Config
-	clientSet *kubernetes.Clientset
+type Executer struct {
+	client.Client
+	log          logr.Logger
+	routines     map[types.UID]*routine
+	routinesLock sync.Mutex
+	runner       runnerFunc
 }
 
-var _ Executer = executer{}
+// runnerFunc is a function that runs the command and returns the stdout, stderr and error
+// it is configurable in Executer for testing purposes
+type runnerFunc func(ctx context.Context, command []string) (string, string, error)
 
-// NewExecuter builds the executer
-func NewExecuter(config *restclient.Config) (Executer, error) {
+// NewExecuter builds the Executer
+func NewExecuter(client client.Client) (*Executer, error) {
 	logger := ctrl.Log.WithName("executer")
 
-	clientSet, err := kubernetes.NewForConfig(config)
-	if err != nil {
-		logger.Error(err, "failed building k8s client")
-		return nil, err
-	}
+	return &Executer{
+		Client:   client,
+		log:      logger,
+		routines: make(map[types.UID]*routine),
+		runner:   run,
+	}, nil
+}
 
-	return &executer{
-		log:       logger,
-		config:    config,
-		clientSet: clientSet,
+// NewExecuter builds an Executer with configurable runnerFunc for testing
+func NewFakeExecuter(client client.Client, fn runnerFunc) (*Executer, error) {
```

Comment:
I think it's a good rule of thumb to separate test code from production code,  based on that I'd expect this method to be  in a different file (and can probably made private as well).

Reply:
good point, I'll move it to the test code